Project 2: Global Value Numbering (#103)
sampsyo/cs6120: content/blog/2019-10-23-global-value-numbering/index.md

title = "Global Value Numbering"
[extra]
bio = """
  [Alexa VanHattum][] is a second-year student interested in the intersection of compilers and formal methods. She also enjoys feminist book clubs and cooking elaborate [fish truck][] meals.

  [Gregory Yauney][] is a second-year student working on machine learning and digital humanities.
  
[alexa vanhattum]: https://www.cs.cornell.edu/~avh
[gregory yauney]: https://www.cs.cornell.edu/~gyauney
[fish truck]: https://www.triphammermarketplace.com/events/
"""
[[extra.authors]]
name = "Alexa VanHattum"
link = "https://www.cs.cornell.edu/~avh"
[[extra.authors]]
name = "Gregory Yauney"
link = "https://www.cs.cornell.edu/~gyauney"
+++

### Overview

If you follow [PL Twitter][pltwitter]™, you may have seen this tweet go by from our very own [Adrian][]:

<img src="twitter.jpg" width="500"/>

What, you may ask, is value numbering, and how is it so mind-blowing? 

The basic premise of [value numbering][vn] is that we can make our code more efficient by removing duplicate computation during compilation.
In particular, if we assign values to every “unique” piece of computation, we can save on work by only actually running that piece once.

> At every subsequent use, we can then refer to the already-computed value.

Author: Added a bit more about this in, while keeping the computation reuse example, thanks!

exit:
  print x;
```

To convert these basic blocks to Bril code, we rename variables and insert a `phi` instruction when the control flow merges again in the last block.

```
start:
  br cond1 left right;
left:
  x1 : int = const 1;
  jmp exit;
right:
  x2 : int = const 2;
  jmp exit;
exit:
  x3 : int = phi x1 x2;
  print x3;
```

Having unique variable assignments is enormously helpful in implementing optimization---most industrial-strength compilers, including [LLVM][] and [GCC][], rely on it internally. We followed a standard algorithm to convert Bril programs to SSA form. 

The first pass inserts `phi` nodes wherever control flow merges multiple variable definitions. 
We determine this condition by computing the [_dominance frontier_][domf] (essentially, boundaries in the control flow graph where control flow can recombine) for every basic block. 
For every block in the dominance frontier of every variable, we insert a `phi` node with as many arguments as there are predecessors to the block. 
One bug we hit with this part of the implementation was realizing that we only should insert `phi` nodes at the dominance frontier for variables that we assigned to multiple times. 

> For convenience of later analysis, we also augmented the base algorithm to track which basic blocks (or source) each `phi` argument originated from (for the above example, our generated `phi` would be `x3 : int = phi x1 x2 left right;`)  

Reviewer: Huh, that's a pretty good idea… maybe I should standardize an "SSA Bril" extension that encodes phi-nodes this way.

The second pass is a recursive renaming algorithm that uses a stack to determine the correct version of a variable name for every point in the program.
The trickiest part of implementing this algorithm was traversing the basic blocks in the correct order.
The high level algorithm specifies that you should recursively traverse the children of each block in the dominator tree. 
However, we found that not only did we ned to traverse the children, but we needed to traverse them _in order_ relative to the control flow graph. 
In particular, Bril’s ecosystem already had a utility for getting the dominators of every block. To get the dominance tree, we calculated the transitive reduction of this relation. 

> Our implementation additionally extending the dominance tree such that every list of children was ordered relative to a reverse post-order traversal of the control flow graph.

Reviewer: ```suggestion
Our implementation additionally extends the dominance tree such that every list of children was ordered relative to a reverse post-order traversal of the control flow graph.
```

…or "extended."

In later testing (using an example, `test/gvn/constant-folding-tau-example.bril`, ported from the [class notes of Prof. Mooly Sagiv of Tel Aviv University)][tau], we found one minor bug with our SSA implementation. 
Bril programs do not have a type checker and are aggressively dynamic: it’s perfectly valid to assign to a variable in one branch of a conditional but not the other. 
The base SSA algorithm does not account for this, so we fail to properly rename in the second step of our algorithm. 

> For now, we work around this case by assigning a default value to to the variable before we branch. 

Reviewer: Yeah, that's totally the right thing to do—to imagine that every variable is assigned to with a special "uninitialized" value at the entry block. I should probably mention this in future SSA lectures.

[ssa]: https://en.wikipedia.org/wiki/Static_single_assignment_form
[llvm]: https://llvm.org/docs/tutorial/OCamlLangImpl7.html
[gcc]: https://gcc.gnu.org/onlinedocs/gccint/SSA.html
[domf]: https://en.wikipedia.org/wiki/Dominator_(graph_theory)
[tau]: http://www.cs.tau.ac.il/~msagiv/courses/pa07/lecture2-notes-update.pdf

## Global value numbering

> There are two main approaches to global value numbering, as described in [Value Numbering][], Briggs, Cooper, and Simpson, 1997, both of which ultimately require input code in SSA form.

Reviewer: Use quotation marks around titles to make it clear that it's the title of something.

Hash-based techniques, like those used in local value numbering, hash operations and find redundant values by comparing to previously hashed operations.
Partitioning algorithms, on the other hand, divide all operations into equivalence classes.

We chose to implement hash-based value numbering because it allows for copy propagation extensions and operates on the dominator tree computed during conversion to SSA.
While converting to SSA and performing global value numbering can be performed together in one pass over the original code due to their similarity, we chose to implement them separately for modularity.
A distinct SSA conversion allows more downstream optimizations to operate on the program's SSA form in addition to allowing testing of both passes independently.

> We implemented the following pseudocode, adapted from Figure 4 of the aforementioned paper, so called because it is a 'dominator-based value numbering technique'.

Reviewer: ```suggestion
We implemented the following pseudocode, adapted from Figure 4 of the aforementioned paper, so called because it is a "dominator-based value numbering technique."
```

In most :us: style guides, double quotes are the top-level thing, and periods go inside the punctuation.

For example, the following Bril:

```
void main() {
entry:
  a : int = const 1;
  b : int = const 2;
  c : int = add a b;
  print c;
}
```

Can be replaced with:
```
void main() {
entry:
  c : int = const 3;
  print c;
}
```

We extend our global value numbering to perform constant folding by keeping a mapping from variable names defined as constants to their integer or boolean literal values.
When we encounter and arithmetic operation, we check whether all operands are constant, and fold (that is, perform the operation and drop in the result) if so.
The argument constants can then often be removed entirely with a later dead code elimination pass.

When implementing constant folding, we [found a bug][bug] in Bril's existing local value numbering implementation.
When Bril programs have division by zero, the local constant folding failed with an exception.
We made the design judgment that compiler passes should not fail on misbehaving code, but instead generate code with the same behavior (for example, the implementation's dynamic error could exist on a branch that never actually executes).

> We modifying this behavior in both the central Bril local value and our new global value numbering to instead bail out on constant folding when encountering a potential exception in the folding step.

Reviewer: ```suggestion
We modified this behavior in both the central Bril local value and our new global value numbering to instead bail out on constant folding when encountering a potential exception in the folding step.
```

> We modifying this behavior in both the central Bril local value and our new global value numbering to instead bail out on constant folding when encountering a potential exception in the folding step.

Reviewer: "Bril local value" -> +numbering?

[bug]: https://github.com/sampsyo/bril/issues/40

<!-- Difficulties:
- Recursively visiting blocks in the dominator tree in reverse post order is not enough to ensure that a phi node's arguments have been processed in the absence of backedges. 
We also had to sort the immediately dominated blocks by their order in the CFG.
- Can't do constant propagation because Bril operations require registers as operands.
- `examples/gvn/constant-folding-calpoly-example.bril`: There's a division by zero when running constant folding, but the compiler shouldn't crash!
We filed an issue with Bril's local value numbering and fixed our implementation.

>  -->

Reviewer: I think you've covered all of this in un-commented text.

```
void main(a0 : int, b0 : int, c0 : int, d0 : int, e0 : int, f0 : int) {
B1:
  u0 : int = add a0 b0;
  v0 : int = add c0 d0;
  w0 : int = add e0 f0;
  cond : bool = const true;
  br cond B2 B3;
B2:
  x0 : int = add c0 d0;
  y0 : int = add c0 d0;
  jmp B4;
B3:
  u1 : int = add a0 b0;
  x1 : int = add e0 f0;
  y1 : int = add e0 f0;
  jmp B4;
B4:
  u2 : int = phi u0 u1 B2 B3;
  x2 : int = phi x0 x1 B2 B3;
  y2 : int = phi y0 y1 B2 B3;
  z0 : int = add u2 y2;
  u3 : int = add a0 b0;
}
```

Our implementation removes the redundant additions in blocks `B2` and `B3`.
It also removes from block `B4`: 1) the meaningless phi node `u2 : int = phi u0 u1 B2 B3;`, which has arguments that are always equal, and 2) the redundant phi node `y2 : int = phi y0 y1 B2 B3;`, which is identical to the phi node that precedes it. 
Here is the full output, which matches the correct output in the paper:

```
void main(a0 : int, b0 : int, c0 : int, d0 : int, e0 : int, f0 : int) {
B1:
  u0: int = add a0 b0;
  v0: int = add c0 d0;
  w0: int = add e0 f0;
  cond: bool = const true;
  br cond B2 B3;
B2:
  jmp B4;
B3:
  jmp B4;
B4:
  x2: int = phi v0 w0 B2 B3;
  z0: int = add u0 x2;
  ret ;
}

> ```

Reviewer: Yay! I really like the idea of reproducing a specific example from the paper.

### Trying to get fewer instructions

We treat the number of instructions as a proxy for code performance, where fewer lines of code is superior.
For each of a number of Bril test programs, we report in the following graph:
1. The number of instructions after converting to SSA, converting out of SSA, and running Bril's existing [trivial dead code elimination][] (TDCE), shown in blue.
2. The number of instructions after converting to SSA, running GVN, converting out of SSA, and running TDCE, shown in red.

<img src="eval-correctness.png" width="500"/>

### Bigger benchmarks with TypeScript frontend and comparison with LVN

> For a slightly more realistic analysis, we ran GVN on the more complex programs afforded by Bril's TypeScript frontend. In particular, none of these programs were written with intentional redundancy like some of our correctness test cases (thought the conversion from TypeScript to Bril does cause some), so they portray a more realistic evaluation of the benefits of value numbering.

Reviewer: ```suggestion
For a slightly more realistic analysis, we ran GVN on the more complex programs afforded by Bril's TypeScript frontend. In particular, none of these programs were written with intentional redundancy like some of our correctness test cases (though the conversion from TypeScript to Bril does cause some), so they portray a more realistic evaluation of the benefits of value numbering.
```

In addition to the existing test programs from project 1, we also implemented a quadratic formula calculation (`test/gvn/quadratic.ts.bril`), [fizz buzz][] (`test/gvn/fizz-buzz.ts.bril`), and a naive [Sieve of Eratosthenes][] without arrays (`test/gvn/check-primes.ts.bril`).
All TypeScript programs were first processed by the `ts2bril` frontend.
We encountered an implementation difficulty: the code generated by the TypeScript frontend sometimes inserts `jmp` instructions after returns in the same basic block.

> We had to manually remove these in order for SSA conversion and dominator tree creation to work.

Reviewer: Hmm, yeah… I guess maybe the thing to do in the future would be to have the CFG generator transform `ret`s into jumps to a common exit block that just contains `ret` all by itself, in the case where there are multiple return points.
